### PR TITLE
make ValidateTransactionWithState also consider sgt balance

### PR DESCRIPTION
**Pull request by blockchaindevsh**
This PR fixes https://github.com/ethstorage/op-geth/issues/11

### CODE REVIEW

File: core/txpool/validation.go [hunk 1]
```diff
@@ -251,30 +252,35 @@ func ValidateTransactionWithState(tx *types.Transaction, signer types.Signer, op
...
+	sgtBalance := new(uint256.Int).SetBytes(sgtBalanceValue[:]).ToBig()
+
 	if opts.L1CostFn != nil {
 		if l1Cost := opts.L1CostFn(tx.RollupCostData()); l1Cost != nil { // add rollup cost
 			cost = cost.Add(cost, l1Cost)
 		}
 	}
-	if balance.Cmp(cost) < 0 {
-		return fmt.Errorf("%w: balance %v, tx cost %v, overshot %v", core.ErrInsufficientFunds, balance, cost, new(big.Int).Sub(cost, balance))
+	if balance.Cmp(cost) < 0 && sgtBalance.Cmp(cost) < 0 {
```
Comment (qizhou) [inline core/txpool/validation.go:272]: The validation logic is not consistent with execution logic where msg.from is 0x0 (https://github.com/ethstorage/op-geth/blob/25669f6dbcda42fa8809094e53f164cc6369748c/core/state_transition.go#L342)

Comment (blockchaindevsh) [reply core/txpool/validation.go:272]: Good catch, it seems the check for zero account is no longer necessary after we avoided evm.call, so removed. 

The UseSoulGasToken flag is added back.

**Review by qizhou — APPROVED** [review]

**Comment by qzhodl** [comment]
> This PR fixes #11

Does this mean that if the sender’s native gas token balance is zero or insufficient to cover the gas fee, the transaction will always fail, regardless of the soul gas token balance?

**Comment by blockchaindevsh** [comment]
> > This PR fixes #11
> 
> Does this mean that if the sender’s native gas token balance is zero or insufficient to cover the gas fee, the transaction will always fail, regardless of the soul gas token balance?

It will only fail when both native balance and sgt balance are not sufficient.

**Comment by qzhodl** [comment]
> > > This PR fixes #11
> > 
> > 
> > Does this mean that if the sender’s native gas token balance is zero or insufficient to cover the gas fee, the transaction will always fail, regardless of the soul gas token balance?
> 
> It will only fail when both native balance and sgt balance are not sufficient.

what I ask is the condition before this PR

**Comment by blockchaindevsh** [comment]
> > > > This PR fixes #11
> > > 
> > > 
> > > Does this mean that if the sender’s native gas token balance is zero or insufficient to cover the gas fee, the transaction will always fail, regardless of the soul gas token balance?
> > 
> > 
> > It will only fail when both native balance and sgt balance are not sufficient.
> 
> what I ask is the condition before this PR

Before this change, it only checks native balance, so yes.

**Review by qzhodl — APPROVED** [review]

**Review by pochenai — APPROVED** [review]

**Pull request merged**